### PR TITLE
Create dockerfiles for kubectl-gadget and CO-RE tools

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Please see individual commits for more information.

--- 

It's still not clear if we want to push these images to a container registry, so this PR doesn't handle it.

### CODE REVIEW

File: .dockerignore
Comment (alban) [inline .dockerignore:10]: Could we use a different dockerignore file for the kubectl-gadget image? Otherwise, the addition of `.git` would invalidate the cache too often for the main image.

How to use a different dockerignore: https://github.com/moby/moby/issues/12886#issuecomment-480575928

File: gadget.core.Dockerfile
Comment (alban) [inline gadget.core.Dockerfile:100]: `gadget.Dockerfile` does not have `-f`. I think the 2 Dockerfiles should be consistent.

**Review by alban — COMMENTED** [review]
> It's still not clear if we want to push these images to a container registry, so this PR doesn't handle it.

What about pushing the images to the same container registry but with the tag prefix 'core-'?

File: .dockerignore
Comment (mauriciovasquezbernal) [reply .dockerignore]: Updated. I left the rules for `Makefile` and `*.mk` in case BuildKit is not available.

File: gadget.core.Dockerfile
Comment (mauriciovasquezbernal) [reply gadget.core.Dockerfile:100]: Updated both of them to have `-f`.

**Comment by mauriciovasquezbernal** [comment]
I'll wait for https://github.com/kinvolk/inspektor-gadget/pull/597 to go in first before updating this PR.

**Comment by mauriciovasquezbernal** [comment]
> I'll wait for #597 to go in first before updating this PR.

Actually I changed my mind. I wanted to tackle https://github.com/kinvolk/inspektor-gadget/issues/625 in this PR too but it seems it's too much, so I propose to merge it as it's.

File: gadget.core.Dockerfile
Comment (eiffel-fl) [inline gadget.core.Dockerfile:67]: Do we forcefully need `bash` there?
I think adding bash to an alpine image increase the image size and so we loose the benefit of using alpine.

File: kubectl-gadget.Dockerfile
Comment (eiffel-fl) [inline kubectl-gadget.Dockerfile:1]: What is the use case of this Dockerfile?
To enable having a portable `kubectl-gadget` where user only needs to give the `.kubeconfig` to interact with an existing cluster?

**Review by eiffel-fl — COMMENTED** [review]
Hi!

Thank you for it, it seems really cool, particularly the image for CO-RE.
I nonetheless have two questions:

Best regards.

File: gadget.core.Dockerfile
Comment (mauriciovasquezbernal) [reply gadget.core.Dockerfile:67]: Yes, we need bash there because our entrypoint is a bash script. Anyway, I don't think adding bash will affect the image size that much.

File: kubectl-gadget.Dockerfile
Comment (mauriciovasquezbernal) [reply kubectl-gadget.Dockerfile:1]: Yes, it's another way to ship kubectl-gadget, if you already have docker or a similar runtime you can just pull the container image and run kubectl-gadget from there without having to download the binary by hand.

File: gadget.core.Dockerfile
Comment (eiffel-fl) [reply gadget.core.Dockerfile:67]: > affect the image size that much.

What is our image size with alpine?
Because `bash` seems to add around 4.5 MB:

```
alpine-bash                             latest                                                   dea7fd1962e5   4 minutes ago   10MB
alpine                                  latest                                                   0ac33e5f5afa   2 weeks ago     5.57MB
```

File: kubectl-gadget.Dockerfile
Comment (alban) [inline kubectl-gadget.Dockerfile:22]: Why not `/bin/kubectl-gadget`?

Is there any documentation how to use that image, or how to extract kubectl-gadget from that image?

**Review by alban — CHANGES_REQUESTED** [review]
Some notes:
* Add comments near the new Makefile target and at the top of the 3 Dockerfiles to explain the difference between them (info you've written in the commit messages)
* The image names will need to be renamed after #597 gets merged (if it gets merged first)

File: kubectl-gadget.Dockerfile
Comment (mauriciovasquezbernal) [reply kubectl-gadget.Dockerfile]: Updated it to be /bin/kubectl-gadget.

**Comment by mauriciovasquezbernal** [comment]
I updated the PR to build the core container image. However it's not pushed nor tested. I think it could be handled in https://github.com/kinvolk/inspektor-gadget/issues/625. Also to avoid more conflicts with https://github.com/kinvolk/inspektor-gadget/pull/597.

File: docs/CONTRIBUTING.md
Comment (marga-kinvolk) [inline docs/CONTRIBUTING.md:153]:
```suggestion
As you can see in `gadget.bcc.Dockerfile`, the gadget container image
```

File: docs/CONTRIBUTING.md
Comment (marga-kinvolk) [inline docs/CONTRIBUTING.md:40]: What about non CO-RE gadgets that are not BCC (traceloop, network-policy advisor)?

File: gadget-container/entrypoint.sh
Comment (marga-kinvolk) [inline gadget-container/entrypoint.sh:111]:
```suggestion
      # In the gadget image, /usr/src is a symlink to /host/usr/src (see Dockerfile).
```

File: kubectl-gadget.Dockerfile
Comment (marga-kinvolk) [inline kubectl-gadget.Dockerfile:1]: Maybe add one more sentence regarding the purpose of this container? Given that the plugin is a statically compiled binary, it's not immediately obvious why someone might want to have a container with it.

File: gadget-container/entrypoint.sh
Comment (marga-kinvolk) [inline gadget-container/entrypoint.sh:62]: Why did you remove the prefix here? There isn't a "Dockerfile" file. Do you mean to point to any of the Dockerfiles? I think maybe gadget.*.Dockerfile would be clearer in that case? Otherwise, if I'm reading this and then find that there is no Dockerfile, I'd assume something got broken in between.

File: gadget-container/entrypoint.sh
Comment (marga-kinvolk) [reply gadget-container/entrypoint.sh:111]: Or maybe gadget.*.Dockerfile, depending on the reply to my comment above.

**Review by marga-kinvolk — COMMENTED** [review]
Some comments about documentation

File: kubectl-gadget.Dockerfile
Comment (blanquicet) [inline kubectl-gadget.Dockerfile:1]: Why you didn't create a make target to build this container?

File: kubectl-gadget.Dockerfile
Comment (blanquicet) [inline kubectl-gadget.Dockerfile:4]: I was expecting it to be used like this:

```bash
$ docker run --rm -it kubectl-gadget trace exec
NODE             NAMESPACE        POD              CONTAINER        PCOMM            PID    PPID   RET ARGS
...
```

In that case we could use scratch as base image.

File: gadget.core.Dockerfile
Comment (blanquicet) [inline gadget.core.Dockerfile:53]: It is missing `rpm2cpio` and `cpio` for OpenShift.

**Review by blanquicet — COMMENTED** [review]
I think it looks like a good starting point. There are some things that need to be improved/missing like the documentation of the kubectl-gadget image but I think we can manage it in future PRs.

File: docs/CONTRIBUTING.md
Comment (mauriciovasquezbernal) [reply docs/CONTRIBUTING.md:40]: Added a comment clarifying what happens with them.

File: gadget-container/entrypoint.sh
Comment (mauriciovasquezbernal) [reply gadget-container/entrypoint.sh:62]: I wanted to make it generic and valid for both Dockerfiles, changed to `gadget.*.Dockerfile`.

File: gadget.core.Dockerfile
Comment (mauriciovasquezbernal) [reply gadget.core.Dockerfile:5]: Makes sense. Reworded it in a similar way to what you suggested.

File: kubectl-gadget.Dockerfile.dockerignore
Comment (mauriciovasquezbernal) [reply kubectl-gadget.Dockerfile.dockerignore:5]: I think it's not needed. I'll remove it.

File: kubectl-gadget.Dockerfile
Comment (mauriciovasquezbernal) [reply kubectl-gadget.Dockerfile:4]: That's one option. Another one is to run `docker run --rm -it kubectl-gadget /bin/sh` and then execute kubectl-gadget from there, it provides lower latency as it doesn't need to create a new container instance for each kubectl-gadget invocation. Alpine is around 6MB, so I'd say it's better to provide an image with some more tools than just providing the kubectl-gadget binary.

File: gadget.core.Dockerfile
Comment (mauriciovasquezbernal) [reply gadget.core.Dockerfile:53]: These are needed for downloading the kernel headers but this image is about CO-Re tools, hence those should not be needed. What happens if they are not installed?, is there a warning or the whole gadget container creation fails?

File: kubectl-gadget.Dockerfile
Comment (mauriciovasquezbernal) [reply kubectl-gadget.Dockerfile:1]: Added.

File: gadget.core.Dockerfile
Comment (blanquicet) [reply gadget.core.Dockerfile]: You are right. Currently, it prints an error but it goes ahead with the container initialization. Once we have CORE implementation for all the gadgets, we can avoid downloading the kernel headers if `CONFIG_DEBUG_INFO_BTF` is set.

**Comment by mauriciovasquezbernal** [comment]
For the time being the Github Actions are only compiling the image but it's not pushed anywhere. We need to understand how to publish those images, I have thought about two different possibilities:

1. Use the same registry but different tags
    - ghcr.io/kinvolk/inspektor-gadget:v0.5.0 -> points to default image 
    - ghcr.io/kinvolk/inspektor-gadget:v0.5.0-core
    - ghcr.io/kinvolk/inspektor-gadget:v0.5.0-bcc
    - ghcr.io/kinvolk/inspektor-gadget:v0.5.0-kubectl-gadget
2. Have different registries for each image (TODO: we can use / or - as name separator depending on the container registry)
   - ghcr.io/kinvolk/inspektor-gadget/kubectl-gadget:v0.5.0
   - ghcr.io/kinvolk/inspektor-gadget-core:v0.5.0
   - ghcr.io/kinvolk/inspektor-gadget-bcc:v0.5.0

Anyway, I think it's something we can tackle in https://github.com/kinvolk/inspektor-gadget/issues/625 and we can move ahead with this PR. Any comments?

File: .github/workflows/inspektor-gadget.yml
Comment (alban) [inline .github/workflows/inspektor-gadget.yml:257]:
```suggestion
      uses: actions/cache@v2
      with:
        path: /tmp/.buildx-cache
        key: ${{ runner.os }}-buildx-gadget-${{ hashFiles('gadget.core.Dockerfile') }}
        restore-keys: |
          ${{ runner.os }}-buildx-gadget-
          ${{ runner.os }}-buildx-
```

We could use different cache prefix for the core image build and the non-core image build, since they are not based on the same image. Otherwise, in case of non-exact match on the key, we risk picking up the wrong one, depending on what is faster to build.

Similarly, the non-core image build could use `${{ runner.os }}-buildx-noncore-${{ hashFiles('gadget.Dockerfile') }}`

See https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#matching-a-cache-key

File: gadget.bcc.Dockerfile
Comment (alban) [inline gadget.bcc.Dockerfile:3]:
```suggestion
# Dockerfile for Inspektor Gadget container image.
# This image contains CO-RE and BCC-based gadgets. Its base image is the
# BCC image. It's the default image that is deployed in Inspektor Gadget.
```

File: docs/CONTRIBUTING.md
Comment (alban) [inline docs/CONTRIBUTING.md:40]: Given that `gadget-bcc` contains both bcc-based gadgets and core-based gadgets, it is a bit confused to call it bcc only. What about calling the two files:
1. `gadget-default`
2. `gadget-coreonly`

(and renaming the Dockerfiles accordingly)

**Review by alban — COMMENTED** [review]
Thanks!

**Comment by eiffel-fl** [comment]
Sorry I did not review it, but from my understanding it looks quite cool.

> We need to understand how to publish those images, I have thought about two different possibilities:

Regarding this, I would prefer the second except I think GHCR does not handle nested registries (`inspektor-gadget/kubectl-gadget`).
Anyway, this PR is about "**[c]reat**[ing] dockerfiles for kubectl-gadget and CO-RE tools", so I think it can be merged and we will address GHCR troubles in future(s?) PR.

File: .github/workflows/inspektor-gadget.yml
Comment (mauriciovasquezbernal) [reply .github/workflows/inspektor-gadget.yml:257]: I'm no t sure I follow. What you're saying is that `restore-keys` should be different for both images to avoid picking up a wrong cache when there is not a match on `key`? 

Will using 

```        
restore-keys: |
          ${{ runner.os }}-core-single-buildx
```

and 

```        
restore-keys: |
          ${{ runner.os }}-bcc-single-buildx
```

fix the issue?

Also, do we really need these `restore-keys`?

File: kubectl-gadget.Dockerfile
Comment (mauriciovasquezbernal) [reply kubectl-gadget.Dockerfile:6]: I just tested and it worked fine. It seems the lack of `etc/nsswitch.conf` doesn't affect it. 

```
sudo docker run --rm -it -v /home/vagrant/.kube/config:/.kube/config kubectl-gadget:latest kubectl-gadget trace exec
```

File: docs/CONTRIBUTING.md
Comment (mauriciovasquezbernal) [reply docs/CONTRIBUTING.md]: Renamed it to be `gadget-default` and `gadget-core`.

File: kubectl-gadget.Dockerfile
Comment (alban) [reply kubectl-gadget.Dockerfile:6]: Thanks for testing!

File: .github/workflows/inspektor-gadget.yml
Comment (alban) [reply .github/workflows/inspektor-gadget.yml]: > What you're saying is that restore-keys should be different for both images to avoid picking up a wrong cache when there is not a match on key?

Yes.

If a PR modifies the Dockerfile in some layer, the `${{ hashFiles(...) }}` will be different, but it is still worth falling back to an older cache from some older PR, but it has to be the same image, otherwise there will not be any Docker layer in common.

I think your suggested `restore-keys` should work.

> Also, do we really need these restore-keys?

As far as I understand, the `restore-keys` is just a fallback if the preferred `key` does not match. Having them will improve the perf if there is no exact match.

Comment (mauriciovasquezbernal) [reply .github/workflows/inspektor-gadget.yml]: I checked the examples at https://github.com/actions/cache/blob/main/examples.md and it seems the restore keys are always used like the main key without `${{ hashFiles('...') }}`. I updated this to follow the same approach.

**Comment by mauriciovasquezbernal** [comment]
Is there anything else I should make to get this in? @marga-kinvolk  @alban @blanquicet @eiffel-fl

**Review by eiffel-fl — APPROVED** [review]
I did not test them but I think it has enough reviews and I trust you so let's merge it.
In case of trouble, we can still address them in future PR.

Just be sure you addressed the requested changes before hitting the merge button.

File: docs/CONTRIBUTING.md
Comment (alban) [inline docs/CONTRIBUTING.md:45]:
```suggestion
  * traceloop and the network policy advisor gadgets aren't included,
    they will be included once
    [#371](https://github.com/kinvolk/inspektor-gadget/issues/371) and
    [#172](https://github.com/kinvolk/inspektor-gadget/issues/172) are
    fixed.
```

**Review by alban — APPROVED** [review]
1 suggestion & LGTM. Thanks!

**Pull request merged**
